File: MATLAB/surface/PlaneSurf.m
classdef PlaneSurf < Surface
    
    properties (GetAccess = 'public', SetAccess = 'private')
        pec         %1x4, plane equation coefficients [a b c d]*[x y z 1]' = 0
    end

    methods
        function obj = PlaneSurf(in1,in2)
            
            obj.pec = zeros(1,4);
            
            %constructor
            if nargin > 1
                N = in1;    %normal vector, 3x1
                pt = in2;   %point, 3x1
                N = N/norm(N); %normalize
                obj.pec(:) = [N; -N'*pt];
            else
                obj.pec(:) = in1;
                N = obj.pec(1:3);
                N = N/norm(N); %normalize
                obj.pec(1:3) = N;

            end
            
        end

        %functions required by the Abstract superclass 'Surface'
        
        function h = surfaceHeight(obj,pts)
            %z = -1/c*[a b d]*[x y 1]'
            
            pts(3,:) = 1;
            h = -1/obj.pec(3) * obj.pec([1 2 4])*pts;
            
        end
        
        function dz = surfaceDz(obj,pts)
            %dz = [a b c d]*[x y z 1]'
            
            pts(4,:) = 1;
            dz = obj.pec*pts;
        end

        function N = surfaceNormal(obj,pts)
            n = size(pts,2);
            N = obj.pec(1:3)'*ones(1,n); %faster than repmat
        end
    end

end

File: src/surface/PlaneSurf.cpp
#include <wmrde/surface/PlaneSurf.h>

//overloaded constructor
PlaneSurf::PlaneSurf(const Real PlaneEqnCoeff[4]) {
	Real norm = normVec3(PlaneEqnCoeff); //first three elements of pec should be unit normal vector
	for (int i=0; i<4; i++)
		pec[i] = PlaneEqnCoeff[i]/norm;
}

PlaneSurf::PlaneSurf(const Vec3 normal, const Vec3 point) {
	copyVec3(normal,pec);
	normalizeVec3(pec);
	pec[3] = -dotVec3(pec,point);
}



//define the pure virtual equations required by the Abstract super class
int PlaneSurf::surfaceHeight(const Vec3 pt, int loc, Real& height) {
	//z = -1/c*[a b d]*[x y 1]'
	height = -(pec[0]*pt[0] + pec[1]*pt[1] + pec[3])/pec[2];

	return 0;
}

int PlaneSurf::surfaceDz(const Vec3 pt, int loc, Real& dz, Vec3 normal) {

	//dz = [a b c d]*[x y z 1]'
	dz = pec[0]*pt[0] + pec[1]*pt[1] + pec[2]*pt[2] + pec[3];

	if (normal != 0)
		copyVec3(pec,normal);

	return 0;
}

int PlaneSurf::surfaceNormal(const Vec3 pt, int loc, Vec3 normal) {
	copyVec3(pec,normal);
	return 0;
}




